### ISSUE TITLE
Error due to feature_dict returned as tuples instead of tensors

### ISSUE THREAD

**Issue by chandlerbing65nm** [HSG-AIML/SSLTransformerRS#6]
I'm trying to train the model using `train_d_swin_backbone.py`, but I'm getting the following error:

```
Traceback (most recent call last):
File "train_d_swin_backbone.py", line 115, in <module>
s = trainer.train(train_loader, val_loader)
File "/workspace/SSLTransformerRS/d_swin_utils.py", line 194, in train
features = torch.cat([feature_dict["s1"], feature_dict["s2"]])
TypeError: expected Tensor as element 0 in argument 0, but got tuple
```
from this code line:

https://github.com/HSG-AIML/SSLTransformerRS/blob/934ca2a0b669a05c922f08dced65f629fa2a32e9/d_swin_utils.py#L194

It looks like the error is occurring because `feature_dict["s1"]` and `feature_dict["s2"]` are being returned as tuples instead of tensors. 

### Steps to reproduce

1. Run the script `python train_d_swin_backbone.py`
2. The error will occur during the training phase when the `torch.cat()` function is called on line 194 of the file `d_swin_utils.py`

### Expected behavior

The model should be trained without any errors.

### Additional information

- I am using Pytorch version: 1.11.0
- I am running the script on a NVIDIA V100

It would be great if you can take a look at this, and let me know if you need any additional information to help troubleshoot the issue.

Thanks!

**Comment by chandlerbing65nm** [comment]
I tried to debug it by asserting that `feature_dict` contains all tensors, here is my code:

```python
error_key = None
for key, value in feature_dict.items():
    try:
        assert type(value) == torch.Tensor, f"{key} has type {type(value)}, expected a Tensor"
    except:
        error_key = key
        raise
if error_key is not None:
    print(f"{error_key} has violated the assert statement")
else:
    features = torch.cat([feature_dict["s1"], feature_dict["s2"]])

```

here's the output:

```
Traceback (most recent call last):
  File "train_d_swin_backbone.py", line 115, in <module>
    s = trainer.train(train_loader, val_loader)
  File "/workspace/SSLTransformerRS/d_swin_utils.py", line 197, in train
    assert type(value) == torch.Tensor, f"{key} has type {type(value)}, expected a Tensor"
AssertionError: s1 has type <class 'tuple'>, expected a Tensor
```

From the output, "s1" is a tuple, but it's expected to be a torch.Tensor.
This means that feature_dict["s1"] is returning a tuple instead of a tensor which could be due to an issue in the code that is populating feature_dict.

**Comment by chandlerbing65nm** [comment]
Now I tried to print the `.type` of each input:

```python
      if torch.isnan(sample["s1"]).any() or torch.isnan(sample["s2"]).any():
          # some s1 scenes in sen12ms are known to have NaNs...
          continue

      s1 = sample["s1"].to(self.args.device)
      s2 = sample["s2"].to(self.args.device)
      print(f"s1 has type {type(s1)}, shape {s1.shape}")
      print(f"s2 has type {type(s2)}, shape {s2.shape}")

      images = {"s1": s1, "s2": s2}
      print(f"images has keys {images.keys()}")
      feature_dict = self.model(images)
      print(f"feature_dict has keys {feature_dict.keys()}")

      error_key = None
      for key, value in feature_dict.items():
          try:
              assert type(value) == torch.Tensor, f"{key} has type {type(value)}, expected a Tensor"
          except:
              error_key = key
              raise
      if error_key is not None:
          print(f"{error_key} has violated the assert statement")
```

and I got this output:

```bash
s1 has type <class 'torch.Tensor'>, shape torch.Size([50, 2, 224, 224])
s2 has type <class 'torch.Tensor'>, shape torch.Size([50, 13, 224, 224])
images has keys dict_keys(['s1', 's2'])
feature_dict has keys dict_keys(['s1', 's2'])
Traceback (most recent call last):
  File "train_d_swin_backbone.py", line 115, in <module>
    s = trainer.train(train_loader, val_loader)
  File "/workspace/SSLTransformerRS/d_swin_utils.py", line 198, in train
    assert type(value) == torch.Tensor, f"{key} has type {type(value)}, expected a Tensor"
AssertionError: s1 has type <class 'tuple'>, expected a Tensor
```

It suggests that the `images` dictionary is being passed to the model correctly, and the `feature_dict` returned by the model also has the correct keys. But the value of `s1` is a tuple, which is causing the `AssertionError`.

It's possible that the issue is in the `self.model` itself. It seems that, the model is returning a tuple instead of a tensor, which is causing the issue.

**Comment by chandlerbing65nm** [comment]
https://github.com/HSG-AIML/SSLTransformerRS/blob/934ca2a0b669a05c922f08dced65f629fa2a32e9/Transformer_SSL/models/swin_transformer.py#L878

The root issue comes from the class referenced above.

```python
class DoubleSwinTransformer(nn.Module):
    def __init__(self, encoder1, encoder2):
        super(DoubleSwinTransformer, self).__init__()

        self.backbone1 = encoder1
        self.backbone2 = encoder2

    def forward(self, x):
        return {
            "s1": self.backbone1.forward_features(x["s1"]),
            "s2": self.backbone2.forward_features(x["s2"]),
        }
```
 the output of `self.backbone1.forward_features(x["s1"])` is a tuple, not a tensor. 

And that class is being called here:
https://github.com/HSG-AIML/SSLTransformerRS/blob/934ca2a0b669a05c922f08dced65f629fa2a32e9/train_d_swin_backbone.py#L106

```python
s1_backbone = build_model(config.model_config)
config.model_config.MODEL.SWIN.IN_CHANS = 13
s2_backbone = build_model(config.model_config)
model = DoubleSwinTransformer(s1_backbone, s2_backbone)
```

So, the root cause is from `build_model(config.model_config)` 

And from [backbone_config.json](https://github.com/HSG-AIML/SSLTransformerRS/blob/main/configs/backbone_config.json), I cant find the culprit.

**Comment by chandlerbing65nm** [comment]
Can anyone help? @damianborth @joellehanna97 @mommermi @scheibenreif

**Comment by scheibenreif** [comment]
Hi @chandlerbing65nm 

thanks for using the repo and tracking down the bug. I'm sorry about the delay in our response! You are right, there was a change in the swin transformer forward pass that changed the return type from tensor to tuple. I pushed a fix to make that behavior optional. Hope it works as expected now! Since you started training, I assume the dataset-structure problem from the other issue has been resolved?

Best,

Linus
